adafruit_mcp230xx/mcp23s17.py

```diff
@@ -0,0 +1,264 @@
+        0 = pull-up disabled).  Note pull-down resistors are NOT supported!
+        """
+        return self._read_u8(_MCP23S17_GPPUB)
+
+    @gppub.setter
+    def gppub(self, val):
+        self._write_u8(_MCP23S17_GPPUB, val)
+
+    def get_pin(self, pin):
+        """Convenience function to create an instance of the DigitalInOut class
+        pointing at the specified pin of this MCP23S17 device.
+        """
+        assert 0 <= pin <= 15
+        return DigitalInOut(pin, self)
+
+    @property
+    def interrupt_configuration(self):
+        """The raw INTCON interrupt control register. The INTCON register
+        controls how the associated pin value is compared for the
+        interrupt-on-change feature. If  a  bit  is  set,  the  corresponding
+        I/O  pin  is  compared against the associated bit in the DEFVAL
+        register. If a bit value is clear, the corresponding I/O pin is
+        compared against the previous value.
+        """
+        return self._read_u16le(_MCP23S17_INTCONA)
+
+    @interrupt_configuration.setter
+    def interrupt_configuration(self, val):
+        self._write_u16le(_MCP23S17_INTCONA, val)
+
+    @property
+    def interrupt_enable(self):
+        """The raw GPINTEN interrupt control register. The GPINTEN register
+        controls the interrupt-on-change feature for each pin. If a bit is
+        set, the corresponding pin is enabled for interrupt-on-change.
+        The DEFVAL and INTCON registers must also be configured if any pins
+        are enabled for interrupt-on-change.
+        """
+        return self._read_u16le(_MCP23S17_GPINTENA)
+
+    @interrupt_enable.setter
+    def interrupt_enable(self, val):
+        self._write_u16le(_MCP23S17_GPINTENA, val)
+
+    @property
+    def default_value(self):
+        """The raw DEFVAL interrupt control register. The default comparison
+        value is configured in the DEFVAL register. If enabled (via GPINTEN
+        and INTCON) to compare against the DEFVAL register, an opposite value
+        on the associated pin will cause an interrupt to occur.
+        """
+        return self._read_u16le(_MCP23S17_DEFVALA)
+
+    @default_value.setter
+    def default_value(self, val):
+        self._write_u16le(_MCP23S17_DEFVALA, val)
+
+    @property
+    def io_control(self):
+        """The raw IOCON configuration register. Bit 1 controls interrupt
+        polarity (1 = active-high, 0 = active-low). Bit 2 is whether irq pin
+        is open drain (1 = open drain, 0 = push-pull). Bit 3 is unused.
+        Bit 4 is whether SDA slew rate is enabled (1 = yes). Bit 5 is if SPI
+        address pointer auto-increments (1 = no). Bit 6 is whether interrupt
+        pins are internally connected (1 = yes). Bit 7 is whether registers
+        are all in one bank (1 = no).
+        """
+        return self._read_u8(_MCP23S17_IOCON)
+
+    @io_control.setter
+    def io_control(self, val):
+        self._write_u8(_MCP23S17_IOCON, val)
```

Comment:
If bit 7 (bank) is set to 1 most of the the _MCP23S17_XXX constants will be wrong including the _MCP23S17_IOCON constant so it can't be unset.

You could either mask out that bit to prevent change and print out a warning or raise an exception. I'm not sure which is the best option.

Reply:
I don't have the SPI chips just yet but I can test with real hardware in about a week or so when I get my PCBs in.
If I am understanding this correctly, then the MCP23017 implementation also has this bug.

Reply:
I think if the 2nd bank is disabled, any register values that are changed for the 2nd bank are not applied if the IOCON register bit 7 is set high because the MCP23017 implementation does the same as the MCP23S17 as far as I know.

Reply:
The issue is how the additional registers of the 16 pin variants are addressed. There are two options, which the BANK bit of the IOCON registers determines. This library is hard coded with addresses for BANK=0. So someone could break things by sending in a value that results in the BANK bit being set to 1. I'd suggest checking the input value and raising an exception if that will happen.